contracts/contracts/strategies/CurveAMOStrategy.sol

```diff
@@ -0,0 +1,702 @@
+        _minWithdrawalAmounts[hardAssetCoinIndex] = _amount;
+        // slither-disable-next-line unused-return
+        curvePool.remove_liquidity(requiredLpTokens, _minWithdrawalAmounts);
+
+        // Burn all the removed OTOKEN and any that was left in the strategy
+        uint256 otokenToBurn = oToken.balanceOf(address(this));
+        IVault(vaultAddress).burnForStrategy(otokenToBurn);
+
+        emit Withdrawal(address(oToken), address(lpToken), otokenToBurn);
+
+        // Transfer hardAsset to the recipient
+        hardAsset.safeTransfer(_recipient, _amount);
+
+        // Ensure solvency of the vault
+        _solvencyAssert();
+    }
+
+    function calcTokenToBurn(uint256 _hardAssetAmount)
+        internal
+        view
+        returns (uint256 lpToBurn)
+    {
+        /* The rate between coins in the pool determines the rate at which pool returns
+         * tokens when doing balanced removal (remove_liquidity call). And by knowing how much hardAsset
+         * we want we can determine how much of OTOKEN we receive by removing liquidity.
+         *
+         * Because we are doing balanced removal we should be making profit when removing liquidity in a
+         * pool tilted to either side.
+         *
+         * Important: A downside is that the Strategist / Governor needs to be
+         * cognisant of not removing too much liquidity. And while the proposal to remove liquidity
+         * is being voted on the pool tilt might change so much that the proposal that has been valid while
+         * created is no longer valid.
+         */
+
+        uint256 poolHardAssetBalance = curvePool
+            .balances(hardAssetCoinIndex)
+            .scaleBy(decimalsOToken, decimalsHardAsset);
+        /* K is multiplied by 1e36 which is used for higher precision calculation of required
+         * pool LP tokens. Without it the end value can have rounding errors up to precision of
+         * 10 digits. This way we move the decimal point by 36 places when doing the calculation
+         * and again by 36 places when we are done with it.
+         */
+        uint256 k = (1e36 * lpToken.totalSupply()) / poolHardAssetBalance;
+        // prettier-ignore
+        // slither-disable-next-line divide-before-multiply
+        uint256 diff = (_hardAssetAmount + 1) * k;
+        lpToBurn = diff / 1e36;
+    }
+
+    /**
+     * @notice Remove all hardAsset and OTOKEN from the Curve pool, burn the OTOKEN,
+     * transfer hardAsset to the Vault contract.
+     */
+    function withdrawAll() external override onlyVaultOrGovernor nonReentrant {
+        uint256 gaugeTokens = gauge.balanceOf(address(this));
+        _lpWithdraw(gaugeTokens);
+
+        // Withdraws are proportional to assets held by 3Pool
+        uint256[] memory minWithdrawAmounts = new uint256[](2);
+
+        // Remove liquidity
+        // slither-disable-next-line unused-return
+        curvePool.remove_liquidity(
+            lpToken.balanceOf(address(this)),
```

Comment:
`withdrawAll` should not revert if there is no assets in the strategy. This is important when there is a security incident and `withdrawAllFromStrategies` is called on the vault.

The previous `_lpWithdraw` will work if there are no LP tokens but this `remove_liquidity` call to the Curve pool will fail if there are no LP tokens.

Reply:
Really nice catch! 
I propose this solution [df605e8](https://github.com/OriginProtocol/origin-dollar/pull/2436/commits/df605e889b95f040c7dbf610e8dc2c004d1d307d), let me know if you prefer something different.

Reply:
The solution looks good and is tested